### PR TITLE
feat: Add Service on Daemonset

### PR DESCRIPTION
**Pull request by achetronic**
**Description:**

This PR cover this discussion (https://github.com/open-telemetry/opentelemetry-helm-charts/issues/749)

```
mode: "daemonset"
serviceOnDaemonset: true

service:
    internalTrafficPolicy: ""
```

As talked, **internalTrafficPolicy** value can be empty, `Cluster` or `Local`. 
On DS deployment is Local by default unless set to Cluster

### CODE REVIEW

File: charts/opentelemetry-collector/values.yaml
Comment (TylerHelmuth) [inline charts/opentelemetry-collector/values.yaml:349]: We can set the default value here to `Cluster` since that is the default k8s value

File: charts/opentelemetry-collector/values.yaml
Comment (TylerHelmuth) [inline charts/opentelemetry-collector/values.yaml:15]: I think since we have to check this value against `mode` in NOTES.txt no matter what I think we can move this under the `service` configuration section.  Please add a check in `NOTES.txt` that will fail the install if serviceOnDaemonset is true and mode != daemonset.

File: charts/opentelemetry-collector/values.yaml
Comment (achetronic) [reply charts/opentelemetry-collector/values.yaml]: By the moment I commented it, because it is computed by the template for daemonsets

I did this, because If I set `Cluster` as default value, then people will set `mode: daemonset`, enable the creation of the service, and won't have to set explicitly `internalTrafficPolicy: Cluster` instead of `Local`

WDYT?

File: charts/opentelemetry-collector/values.yaml
Comment (achetronic) [reply charts/opentelemetry-collector/values.yaml]: Fixed! :)

Comment (achetronic) [reply charts/opentelemetry-collector/values.yaml]: fixed! :)

File: charts/opentelemetry-collector/values.yaml
Comment (TylerHelmuth) [reply charts/opentelemetry-collector/values.yaml]: Oh right. Let's allow this value in the schema then but not set it to anything, even empty string, by default. We can leave it commented out in the values.yaml.

Comment (achetronic) [reply charts/opentelemetry-collector/values.yaml]: Hey, I already did it [in this change ](https://github.com/open-telemetry/opentelemetry-helm-charts/pull/755/files#diff-eba1f4eaa1181faf6ab17062aea0b2a92c389a268983766d94eb72f5ec8caa2fR350)

So I think it is ready to go :)

File: charts/opentelemetry-collector/templates/service.yaml
Comment (TylerHelmuth) [inline charts/opentelemetry-collector/templates/service.yaml:34]: Let's move all this logic into the helper.

File: charts/opentelemetry-collector/templates/NOTES.txt
Comment (TylerHelmuth) [inline charts/opentelemetry-collector/templates/NOTES.txt:48]: Please also add a warning is mode is daemonset and the mode is set to cluster.

File: charts/opentelemetry-collector/templates/service.yaml
Comment (achetronic) [reply charts/opentelemetry-collector/templates/service.yaml]: Done! :)

File: charts/opentelemetry-collector/templates/NOTES.txt
Comment (achetronic) [reply charts/opentelemetry-collector/templates/NOTES.txt]: Done! :)

**Comment by achetronic** [comment]
@TylerHelmuth Everything solved :)

File: charts/opentelemetry-collector/Chart.yaml
Comment (achetronic) [reply charts/opentelemetry-collector/Chart.yaml:3]: @povilasv Hey, up to you to decide. I thought it was a minor due to we included a new feature to deploy the service on daemonsets, but if it is considered a patch, please, ping me again and I will change it :)

File: charts/opentelemetry-collector/templates/_helpers.tpl
Comment (TylerHelmuth) [inline charts/opentelemetry-collector/templates/_helpers.tpl:150]: Since the functionality is specifically for mode=daemonset  serviceForDaemonset=true lets switch the condition to check for that. This will make it easier to add a new mode in the future.

File: charts/opentelemetry-collector/templates/_helpers.tpl
Comment (TylerHelmuth) [inline charts/opentelemetry-collector/templates/_helpers.tpl:153]: Instead of another template are we able to use `default`?

```yaml
{{- print (.Values.service.internalTrafficPolicy | default "Local") -}}
```

File: charts/opentelemetry-collector/templates/NOTES.txt
Comment (TylerHelmuth) [inline charts/opentelemetry-collector/templates/NOTES.txt:49]:
```suggestion
[WARNING] Setting internalTrafficPolicy to 'Cluster' on Daemonset is not recommended. Consider using 'Local' instead.
```

File: charts/opentelemetry-collector/templates/_helpers.tpl
Comment (achetronic) [reply charts/opentelemetry-collector/templates/_helpers.tpl]: Fixed! :)

File: charts/opentelemetry-collector/templates/_helpers.tpl
Comment (achetronic) [reply charts/opentelemetry-collector/templates/_helpers.tpl]: Fixed! :)

**Review by TylerHelmuth — APPROVED** [review]
@open-telemetry/helm-approvers please review, I'd like everyone's opinion on this one.

**Comment by achetronic** [comment]
Hello there @TylerHelmuth 
What about this PR? :)

**Comment by TylerHelmuth** [comment]
@achetronic although I have approved, I would like the other approvers/maintainers to give their approvals since this is a pretty impactful change for the collector chart and how we view our responsibility to guide users to certain choices. Feel free to hold off addressing merge conflicts until others have merged, bc they will keep occurring as other PRs get merged in.  

@dmitryax @Allex1 @puckpuck please review.

**Comment by achetronic** [comment]
> @achetronic although I have approved, I would like the other approvers/maintainers to give their approvals since this is a pretty impactful change for the collector chart and how we view our responsibility to guide users to certain choices. Feel free to hold off addressing merge conflicts until others have merged, bc they will keep occurring as other PRs get merged in.
> 
> @dmitryax @Allex1 @puckpuck please review.

Hi @TylerHelmuth thank you for your response :)
Then I will wait for the team to review and discuss, and will solve the conflicts after it 🙏🏼

**Review by Allex1 — APPROVED** [review]

**Comment by antonioberben** [comment]
Can I cry for the merge? 😄

**Review by puckpuck — APPROVED** [review]
🚢 it

**Review by dmitryax — APPROVED** [review]

File: charts/opentelemetry-collector/values.yaml
Comment (dmitryax) [inline charts/opentelemetry-collector/values.yaml:343]: Should we add `enabled` option here which will be false by default for `daemonset`, otherwise `true`?

You may ask why users would need deployment/statufulset without a service? One use case: 1-replica k8scluster receiver, which we already have a preset for. Another use case: a scraping receiver with one endpoint.

@open-telemetry/helm-approvers WDYT?

**Review by dmitryax — COMMENTED** [review]
Approved to soon

File: charts/opentelemetry-collector/values.yaml
Comment (TylerHelmuth) [reply charts/opentelemetry-collector/values.yaml:343]: @dmitryax how were you thinkin we set it to false by default for `mode=daemonset` only?  Are you thinkin the value is a template and `.Values.service.enabled` should always be used via a `tpl` function?

Comment (dmitryax) [reply charts/opentelemetry-collector/values.yaml:343]: We can do it the same way as we currently compute `internalTrafficPolicy`, right?

Comment (TylerHelmuth) [reply charts/opentelemetry-collector/values.yaml:343]: I see what you mean.  Allow `enabled` as a field but don't explicitly set it.  If it is set use it, otherwise use mode to determine the default value.  Ya I like that and it keeps the section consistent with other sections in our values.yaml.  @achetronic can you make that change?

Comment (achetronic) [reply charts/opentelemetry-collector/values.yaml:343]: Let's do it, probably on Monday I will upload the changes 🙏🏼

Comment (achetronic) [reply charts/opentelemetry-collector/values.yaml]: @TylerHelmuth @dmitryax Hello there, sorry for the delay but now the approach is adapted to the discussed. Could you please review the changes? 😄

**Comment by achetronic** [comment]
@TylerHelmuth @dmitryax Hello there, sorry for the delay but now the approach is adapted to the discussed. Could you please review the changes? 😄

**Review by TylerHelmuth — APPROVED** [review]

**Review by dmitryax — APPROVED** [review]

**Comment by dmitryax** [comment]
Thank you, @achetronic

**Pull request merged**
